rosetta/docs/PGLE.md

```diff
@@ -0,0 +1,89 @@
+
+3. **Performance Run**: Finally we have the performance run and this time we make sure we have set the above 3 XLA flags to be true. In addition to that we also provide another XLA flag with the pbtxt file.
+
+```
+export XLA_FLAGS="--xla_gpu_enable_latency_hiding_scheduler=true
+--xla_gpu_enable_async_all_gather=true
+--xla_gpu_enable_async_reduce_scatter=true
+--xla_gpu_pgle_profile_file_or_directory_path=path\to\generated\pbtxt"
+```
+With that we should see good overlap between the computations and collectives.
+
+Optionally, we can set the following environment variables to ensure that the latency hiding scheduler is indeed using the profiling costs to schedule:
+```
+export TF_CPP_VMODULE=profile_guided_latency_estimator=10
+export TF_CPP_MIN_LOG_LEVEL=0
+export TF_CPP_MAX_LOG_LEVEL=100
+```
+This should print something like this:
+```
+PGLE found async wrapped instruction: custom-call.4.1 in custom-call-start.1
+PGLE found latency for async op custom-call-start.1 and (assumed)custom-call-done.1 in instruction costs
+```
+
+### Recommended XLA Flags
+
+In order to get the best performance with PGLE, here is a list of all recommended XLA flags:
+```
+export XLA_FLAGS="--xla_gpu_enable_latency_hiding_scheduler=true
+--xla_gpu_enable_async_all_gather=true
+--xla_gpu_enable_async_reduce_scatter=true
+--xla_gpu_enable_triton_gemm=false
+--xla_gpu_simplify_all_fp_conversions
+--xla_gpu_graph_level=0
+--xla_gpu_enable_async_all_reduce=true
+--xla_gpu_enable_highest_priority_async_stream=true
+--xla_gpu_all_reduce_combine_threshold_bytes=1073741824
+--xla_gpu_all_gather_combine_threshold_bytes=1073741824
+--xla_gpu_reduce_scatter_combine_threshold_bytes=1073741824
+--xla_gpu_enable_pipelined_all_gather=true
+--xla_gpu_enable_pipelined_reduce_scatter=true
+--xla_gpu_enable_pipelined_all_reduce=true
+--xla_gpu_enable_while_loop_double_buffering=true
+--xla_gpu_enable_triton_softmax_fusion=false
+--xla_gpu_enable_all_gather_combine_by_dim=false
+--xla_gpu_enable_reduce_scatter_combine_by_dim=false
+--xla_disable_hlo_passes=rematerialization
+--xla_gpu_pgle_profile_file_or_directory_path=path\to\generated\pbtxt"
+```
+### About Combine thresholds
+
+One last thing to add is regarding the **"combine thresholds"**. Ideally, a higher combining threshold for all-gather, reduce-scatter kernels will ensure the best use of the bandwidth. However, they might also incur some dependencies as XLA would try to combine several async collectives and we might see degradation in overall overlap. On the other hand, using very small combining thresholds would create more number of individual collective ops. This makes overlap easier but might under-utilize the bandwidth. For the best performance, these values may need to be tuned based on the individual model and number of devices used for training.
```

Comment:
at the end add something like: `and hardware` the differnt kind of GPUs and networking gear used can influance that.